codec/src/main/java/io/netty/handler/codec/compression/BrotliEncoder.java

```diff
@@ -47,45 +72,73 @@ public BrotliEncoder() {
     /**
      * Create a new {@link BrotliEncoder} Instance
      *
-     * @param parameters {@link Encoder.Parameters} Instance
+     * @param brotliOptions {@link BrotliOptions} to use.
      */
-    public BrotliEncoder(Encoder.Parameters parameters) {
-        this.parameters = ObjectUtil.checkNotNull(parameters, "Parameters");
+    public BrotliEncoder(BrotliOptions brotliOptions) {
+        this(brotliOptions.parameters());
     }
 
     /**
      * Create a new {@link BrotliEncoder} Instance
      *
-     * @param brotliOptions {@link BrotliOptions} to use.
+     * @param parameters {@link Encoder.Parameters} Instance
      */
-    public BrotliEncoder(BrotliOptions brotliOptions) {
-        this(brotliOptions.parameters());
+    public BrotliEncoder(Encoder.Parameters parameters) {
+        this.parameters = ObjectUtil.checkNotNull(parameters, "Parameters");
     }
 
     @Override
-    protected void encode(ChannelHandlerContext ctx, ByteBuf msg, ByteBuf out) throws Exception {
-        // NO-OP
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        pluggableBuffer = new PluggableByteBufOutputStream();
+        brotliEncoderChannel = new BrotliEncoderChannel(Channels.newChannel(pluggableBuffer), parameters);
+        super.handlerAdded(ctx);
     }
 
     @Override
-    protected ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) throws Exception {
-        // If ByteBuf is unreadable, then return EMPTY_BUFFER.
+    protected void encode(ChannelHandlerContext ctx, ByteBuf msg, ByteBuf out) throws Exception {
         if (!msg.isReadable()) {
-            return Unpooled.EMPTY_BUFFER;
+            return;
+        }
+
+        if (isClosed) {
+            out.writeBytes(msg);
+            return;
         }
 
+        // Plug-in Buffer
+        pluggableBuffer.plugBuffer(out);
+
+        // Compress data and flush it into Buffer
+        brotliEncoderChannel.write(msg.nioBuffer());
+        FLUSH_METHOD.invoke(brotliEncoderChannel);
```

Comment:
why cant we get a release with the method exposed ? Using reflection sounds just not right.

Reply:
That's on my list but I want to use reflection for now so we can ship the patch faster in SNAPSHOTS.